### ISSUE TITLE
HLS: stream does not work on Android 4.1

### ISSUE THREAD

**Issue by DrySs** [flowplayer/flowplayer#812]
Hi there,

I've noticed that on Samsung Galaxy S3 Mini (Android 4.1), HLS does not work with FlowPlayer: audio is playing but not video (infinite FlowPlayer loading screen).

Here is two test pages:

- simple HTML5 video tag: http://dev.dryss.com/signedcookies/
- FlowPlayer 6 : http://dev.dryss.com/signedcookiesfp/

This is the same m3u8 file in both examples, served by a CloudFront distribution with Signed Cookies.

With simple video tag, it's working fine.
With FlowPlayer 6, I have infinite loading screen.

**Comment by nnarhinen** [comment]
Do you see the same behavior on http://demos.flowplayer.org/basics/hls.html ?

On my Samsung Galaxy S3 (Android 4.3) both your examples work just fine.

Because of Android fragmentation we can't have devices for  all possible vendor and android versions combinations to test with but we are happy to accept pull requests against the dev branch if you manage to debug and fix the issue you are seeing.

**Comment by DrySs** [comment]
I'm talking about Samsung Galaxy S3 Mini, which runs under Android 4.1 (and not 4.3).
Do you have any device runinng on Android 4.1?

Your example http://demos.flowplayer.org/basics/hls.html does not work on my 4.1 device.
I cannot (and I don't know) how to proceed to find and fix this issue as this is not my job :p

But what I can say is that jwPlayer works fine on my Android 4.1:
http://www.jwplayer.com/html5/hls/

And standard HTML5 video tag (without any plugin) works fine too:
http://dev.dryss.com/signedcookies/

So, this is obviously a FlowPlayer bug, do you plan to fix it? :)

**Comment by nnarhinen** [comment]
No, I do not have any devices running 4.1. @bbbo do you possess any 4.1 devices?

I can't fix this for every single Android device out in the wild, the combinations are countless.

We will release a fix if we find a fix. Fixing something for a legacy Android version without a device for reproducing the chances of finding a fix are very thin. Not a high priority thing either as it's only one device/version combination not working.

Does your device play a normal VOD stream then: http://demos.flowplayer.org/basics/minimal.html ?

**Comment by DrySs** [comment]
http://demos.flowplayer.org/basics/minimal.html works fine without any problem.

Notice that for HLS stream, I only have audio stream, but loading screen of FlowPlayer as video stream.

**Comment by bbbo** [comment]
will try to find someone with an S3 mini. Samsung updated all my devices 
to at least 4.2 by now.

**Comment by DrySs** [comment]
How should I do to find the bug source with an Android device?

**Comment by nnarhinen** [comment]
@DrySs can you test this page with your S3 Mini: http://jsfiddle.net/90Ld7t98/1/

You should see (or actually I guess you don't see) something like this. If you don't see the "loadeddata fired" text below the video player then the browser if your S3 mini does not send necessary events. 

![2015-05-28 09 29 35](https://cloud.githubusercontent.com/assets/265438/7856917/4d53ac3c-0535-11e5-958d-4b522c4f9ae4.png)

**Comment by DrySs** [comment]
@nnarhinen 

Just tryed, I well have the message "loadedata fired" printed on my screen.

**Comment by nnarhinen** [comment]
Ok, then I'm out of ideas why flowplayer doesn't fire the ready event..

**Comment by DrySs** [comment]
Because when ready event is fired, then loading image is hidden? Video is playing as I can hear the audio. It's the video screen which seems hidden by loading image.

**Comment by nnarhinen** [comment]
Yes, the symptoms seem like ready event would not be fired (video tag is hidden by CSS and loading dots appear)

**Comment by nnarhinen** [comment]
@DrySs  Can you give me a screenshot of your S3 Mini running this fiddle: http://jsfiddle.net/90Ld7t98/3/

**Comment by DrySs** [comment]
Here you are @nnarhinen :

![screenshot_2015-05-29-10-58-53](https://cloud.githubusercontent.com/assets/230791/7879672/b996b7c8-05f2-11e5-9ca9-978ee6052019.png)

**Comment by nnarhinen** [comment]
And if you start playback, do you get a third row with durationchange fired, this time with duration something else than zero?

**Comment by DrySs** [comment]
No, nothing happens: I click play (button transform to stop), then after 2/3sc, it displays back the play button.

durationchange stay at 0.

**Comment by nnarhinen** [comment]
So the HLS stream does not play at all in a plain video tag? Then there's not much flowplayer can do about it.

**Comment by DrySs** [comment]
Yes it does, as I mentionned it in my first message, I have same HLS stream in HTML5 video tag:

http://dev.dryss.com/signedcookies/

It's working very fine on same device.
This IS a FP bug.

**Comment by nnarhinen** [comment]
You just said nothing happens when you click play on my jsfiddle -> your browser does not support playing the HLS stream. Please replicate the jsfiddle setup with an HLS stream that plays on your device so that I get reliable test data. (I can't use your HLS streams as they seem to be doing some cookie magic)

**Comment by DrySs** [comment]
I cannot replicate this HLS stream as it is CloudFront content protected, which need cookie domain to work. It will not works on jsfiddle.

**Comment by nnarhinen** [comment]
You can copy the jsfiddle code to your server.

**Comment by DrySs** [comment]
Only JS part?

**Comment by nnarhinen** [comment]
So that the demo works like it does on jsfiddle.

**Comment by DrySs** [comment]
Done here: http://dev.dryss.com/signedcookies/ with plain video tag.
But now, video won't play anymore.

Following code break the player:

```
                video.addEventListener('loadeddata', function() {
                    log('loaddeddata fired, duration: ' + video.duration);
                });
                video.addEventListener('durationchange', function() {
                    log('durationchange fired, duration: ' + video.duration);
                });
```

**Comment by DrySs** [comment]
Ok finally after waiting like 20sc, video is playing.
But logs stay at 0 for duration.

**Comment by nnarhinen** [comment]
So no third row with "durationchange fired, duration: 139.76" ?

**Comment by DrySs** [comment]
Nop, stays at 0 all video playing.

**Comment by nnarhinen** [comment]
![2015-05-29 13 12 54](https://cloud.githubusercontent.com/assets/265438/7883239/a3ec231a-061d-11e5-83af-4b3db13adc9d.png)

**Comment by DrySs** [comment]
No, I don't have the third log, only 2 first.

**Comment by nnarhinen** [comment]
Please add this code:

```js
video.addEventListener('timeupdate', function() {
    log('timeupdate fired, duration: ' + video.duration);
});
```

Then after you start playback, does it start logging timeupdate events and do those events have the video duration?

**Comment by DrySs** [comment]
Done.

I have "timeupdate fired, duration: 0" printed all over the video is playing.
Still not real video duration.

**Comment by nnarhinen** [comment]
You said JWPlayer and plain video tag work, do they give the duration somewhere?

Flowplayer does not (yet) support durationless setups other than for live setups where you have to configure `live: true`.

**Comment by DrySs** [comment]
I tested jwPlayer with their demo HLS stream, but I did not test with my-self integration (signed cookie HLS stream) because you have to pay for their HLS feature.

I don't know about their duration.

I'm not sure to understand what you said: is that a FP feature leak or a S3 mini bug?

**Comment by nnarhinen** [comment]
Your browser is broken: it does not report duration for video. So there is no way for Flowplayer to know how long the video is and it will break the timeline and time displays.

You can workaround this with configuring flowplayer to think this is a live-video by configuring `live: true`, and then flowplayer will hide all the elements that do not make any sense in live videos (timeline, duration etc)

**Comment by DrySs** [comment]
Yes but I'm not in live configuration, it's pure VOD service, and I need timeline, duration, etc.
What should I do? Force MP4 and disable HLS for Android 4.1?

**Comment by nnarhinen** [comment]
It's not necessarily Android 4.1 that is broken, it might be just this specific samsung configuration. Android fragmentation is a PITA

**Comment by DrySs** [comment]
Problem is that if device does have video duration, video does not work, and there is no fallback to MP4...

**Comment by DrySs** [comment]
Could'nt you add a rule which say if there is no video duration, then fallback to other source (MP4)?

**Comment by nnarhinen** [comment]
Can you try setting http://dev.dryss.com/signedcookiesfp/ to use 

http://cdn.819f.flowplayer.me/adhoc/flowplayer.min.js and then try with your S3 device

**Comment by DrySs** [comment]
Yes it works!!! :+1: 

Very good news, how did you manage to do this?

**Comment by nnarhinen** [comment]
There's some magic applied to situations where durationchange event never fires but playback starts. Then I make flowplayer believe it's in a live setup and it hides duration elements and timeline. It of course discards some of the functionality but at least the video plays.

**Comment by nnarhinen** [comment]
This fix will be available in 6.0.3
